src/hotspot/share/opto/addnode.cpp

```diff
@@ -395,9 +395,192 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
     }
   }
 
+  // Convert a + a + ... + a into a*n
+  Node* serial_additions = convert_serial_additions(phase, can_reshape, bt);
+  if (serial_additions != nullptr) {
+    return serial_additions;
+  }
+
   return AddNode::Ideal(phase, can_reshape);
 }
 
+// Convert a + a + ... + a into a*n
+Node* AddNode::convert_serial_additions(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+  if (is_optimized_multiplication()) {
+    return nullptr;
+  }
+
+  Node* in1 = in(1);
+  Node* in2 = in(2);
+  int in1_op = in1->Opcode();
+
+  bool matched = false;
+  jlong multiplier;
+
+  // Convert (a + a) + a to 3 * a
+  // Look for LHS pattern: AddNode(a, a)
+  if (in1_op == Op_Add(bt) && in1->in(1) == in1->in(2)) {
```

Comment:
It seems each of the if blocks in this method could be its own method that returns true and `multiplier` (passed by reference, I suppose) if pattern matching succeeds.

Reply:
Refactored to do so. Thanks for the input!